Remove GC from odspCache implementation.

NicholasCouri opened the pull request:
Fix #4500 
@markfields @andre4i   this is the change I was planning to remove the GarbageCollector from the odspCache. It depends on the getEntries method being added to the PromiseCache implementation.

The idea is to remove the local cache + GarbageCollector by replacing it with a PromiseCache.

NicholasCouri committed:
Remove GC from odspCache implementation.

markfields commented on packages/drivers/odsp-driver/src/odspCache.ts:
```diff
@@ -2,7 +2,7 @@
  * Licensed under the MIT License.
  */
-
+import { strict as assert } from "assert";
```
Please use `assert` from common-utils to avoid bundle size increases.  Would be nice if we had a lint rule to block this or something...

NicholasCouri replied:
Good to know! Thanks

markfields commented on packages/drivers/odsp-driver/src/odspCache.ts:
```diff
@@ -21,79 +21,46 @@ export interface IPersistedFileCache {
-            this.cache.delete(cachekey);
-        });
+        if (typeof this.pc.getEntries === "function" &&
+            this.pc.getEntries !== null)
```
btw - `typeof null === "object"` so this check is redundant

NicholasCouri replied:
oppss cool!

markfields commented on packages/drivers/odsp-driver/src/odspCache.ts:
```diff
@@ -21,79 +21,46 @@ export interface IPersistedFileCache {
 
-    public constructor(private readonly snapshotExpiryPolicy = 30 * 1000) {}
+    public constructor(private readonly snapshotExpiryPolicy = 30 * 1000) {
+        this.pc = new PromiseCache<string, any>({ expiry: {policy:"sliding", durationMs: this.snapshotExpiryPolicy} });
```
This is tricky.  I don't see a way to preserve existing timeout semantics.  With sliding policy, PromiseCache resets the timeout on read as well as write.  With absolute policy, PromiseCache never resets the policy, not even on write.  But the existing behavior is to reset the timeout on write only.

NicholasCouri replied:
I see - What if I change to absolute ? It would not update on read. Isn't it ?

markfields replied:
Unless I'm misreading the code, for Absolute policy it _never_ resets the timers, not even on write.

We could discuss adding a new policy type, think it over and see how that would work out.

NicholasCouri replied:
Yes, perhaps we could add a new policy type that would simply call 
        this.gc.cancel(key);
        this.gc.schedule(key, this.snapshotExpiryPolicy);

When the entry is being updated. 
Also, so I can understand it better, do we need to preserve the existing semantics ? When setting to sliding, we do cancel/schedule during gets. Why wasn't it the design ?  Do you know ?

andre4i commented on packages/drivers/odsp-driver/src/odspCache.ts:
```diff
@@ -21,79 +21,46 @@ export interface IPersistedFileCache {
 
-    public constructor(private readonly snapshotExpiryPolicy = 30 * 1000) {}
+    public constructor(private readonly snapshotExpiryPolicy = 30 * 1000) {
+        this.pc = new PromiseCache<string, any>({ expiry: {policy:"sliding", durationMs: this.snapshotExpiryPolicy} });
```
there's a minor issue with the spaces and formatting of this json. VSCode should take care of that for you

NicholasCouri replied:
I might be missing some extension as it did not fix it - will double check.

markfields replied:
@andre4i - I also don't have an extension for this, what's your secret?

andre4i replied:
Not sure it's because of an extension. The only extra thing I have installed is `eslint` but it's probably configured poorly as it doesn't match the project's lint rules. This should work out of the box for javascript/ts/json: https://code.visualstudio.com/docs/editor/codebasics#_formatting

markfields commented on packages/drivers/odsp-driver/src/odspCache.ts:
```diff
@@ -21,79 +21,46 @@ export interface IPersistedFileCache {
+        if (typeof this.pc.getEntries === "function" &&
+            this.pc.getEntries !== null)
+        {
+            this.pc.getEntries()
```
This will give you an array full of `Promise`s.  In order to inspect the data you'll need to await them, which I don't see you doing.

NicholasCouri replied:
I think we don't need to wait for them as we are just inspecting the keys. Isn't it ?

markfields replied:
ohhhh right

markfields reviewed (commented):
I know this was my idea when we were brainstorming, but I have some concerns with it now that I see it spelled out.  See some of my comments.

I have an idea for another approach to the bug which I'll poke around at for a few minutes and share with you @NicholasCouri if it seems promising (har har).

[update] - take a look at #9420 as an alternative

NicholasCouri committed:
Addressing initial comments.

NicholasCouri committed:
Reverting back from absolute to sliding.

NicholasCouri commented:
We decided to simply fix the code and manually remove the GarbageCollector implementation without consuming anything from the commutils.

NicholasCouri committed:
Removing dependency from PromiseCache and changing
the Expiration cache to have a single entry per
documentId.
Removing 2 secs timeout from the UTs and starting to
rely on the cleanup at the end of the test to remove
the cache entries.

NicholasCouri committed:
Adjusting spacing.

markfields commented on packages/drivers/odsp-driver/src/odspCache.ts:
```diff
@@ -77,30 +40,52 @@ export class LocalPersistentCache implements IPersistedCache {
+        this.removeDocIdEntriesFromCache(file.docId);
+    }
+
+    private removeDocIdEntriesFromCache(docId: string) {
```
Do you also want to do removeExpirationEntry(docId) in here?

markfields commented:
@vladsud - What if we just got rid of this default implementation and require hosts to pass one in?  And move this to some test package.  I guess this is a nice-to-have for ramping up when someone's first building out a host, but I would assume everyone trying to develop a production-level solution would roll their own with better capabilities than a simple Map provides.

I'm wondering because we're in this slightly awkward position of owning and maintaining this code, but otherwise usually taking a stance that "the host owns the cache policy".  And we could save a few bytes in our bundle size :)

Just a thought.

vladsud commented:
I'd not mind, but worth pointing out that there are two reasons cache is useful:
1. General perf improvements, including boot perf.
   - to me this is optional, and default cache does not solve it anyway.
2. Reducing number of calls to SPO when summarizer starts (reusing snapshot from main client)
   - this one is more nuanced and that's where this cache helps with.

I think it's fine to remove it, but we need to clearly articulate impact to users in the form of breaking.md and comments around driver factory.

NicholasCouri committed:
Addressing review comments.

markfields commented on packages/drivers/odsp-driver/src/test/epochTests.spec.ts:
```diff
@@ -17,7 +17,7 @@ import { getHashedDocumentId } from "../odspPublicUtils";
 import { mockFetchOk, mockFetchSingle, createResponse } from "./mockFetch";
 
-const createUtLocalCache = () => new LocalPersistentCache(2000);
+const createUtLocalCache = () => new LocalPersistentCache();
```
I wonder if it would be better or worse to just create the cache once and reuse it in all the tests.  It would let us actually exercise the capability to manage different documents' entries independently.

Eh, except the same docId is used every time, so slightly less interesting?  Just thinking out loud, probably shouldn't change it.

NicholasCouri replied:
yes, not sure if it is worthy it.

markfields reviewed (approved):
🎂 LGTM!  Although you should consider adding UTs for `LocalPersistedCache` directly, to test it directly, and guard against any typos during refactoring.

NicholasCouri commented:
Hi Mark, I just realized the comments regarding the UTs - the EpochTracker does manipulate and invoke all the methods from the LocalPersistentCache. The mapping of the code straight-forward.
If you still think we need it, I can easily file a small bug to add a new UT and address the concern.


________________________________
From: Mark Fields ***@***.***>
Sent: Tuesday, March 15, 2022 4:45 PM
To: microsoft/FluidFramework ***@***.***>
Cc: Nicholas Couri ***@***.***>; State change ***@***.***>
Subject: Re: [microsoft/FluidFramework] Remove GC from odspCache implementation. (PR #9415)


@markfields approved this pull request.

🎂 LGTM! Although you should consider adding UTs for LocalPersistedCache directly, to test it directly, and guard against any typos during refactoring.

—
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2FFluidFramework%2Fpull%2F9415%23pullrequestreview-910944541&data=04%7C01%7Cnichoc%40microsoft.com%7C7154b6b2669a4f9da57908da06ddd7f5%7C72f988bf86f141af91ab2d7cd011db47%7C0%7C0%7C637829847141515677%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=siI7DZ81a5pzOQD%2FFjAWAWixsGtBKXZmCHvcLpH6eJo%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAGIMYCGD4V7LR6ESTEDO6JLVAEOINANCNFSM5QKK26RQ&data=04%7C01%7Cnichoc%40microsoft.com%7C7154b6b2669a4f9da57908da06ddd7f5%7C72f988bf86f141af91ab2d7cd011db47%7C0%7C0%7C637829847141515677%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=xxTrAzwTgJXt2uVsLiTUadfi9mxseNRcQ3YAEiiuztA%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Cnichoc%40microsoft.com%7C7154b6b2669a4f9da57908da06ddd7f5%7C72f988bf86f141af91ab2d7cd011db47%7C0%7C0%7C637829847141515677%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=%2F4%2BIiOyoAzzhT%2BW2dMAUB7OSn9Q4zOutaD6XFsTNl2w%3D&reserved=0> or Android<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Cnichoc%40microsoft.com%7C7154b6b2669a4f9da57908da06ddd7f5%7C72f988bf86f141af91ab2d7cd011db47%7C0%7C0%7C637829847141565648%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=mnxdnooa2742nlacanFHrL7ECrC5s0fAFz6JDGni6oE%3D&reserved=0>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>
